### PR TITLE
fix: panic with datafusion 21

### PR DESCRIPTION
**Pull request by hengfeiyang**
- [x] update Datafusion to v21.1.0
- [x] fix panic when handling Datafusion error

### CODE REVIEW

**Review by oasisk — APPROVED** [review]

**Review by prabhatsharma — APPROVED** [review]

**Comment by hengfeiyang** [comment]
fixed #494

File: Cargo.lock [hunk 1]
```diff
@@ -1712,9 +1668,9 @@ dependencies = [
 
 [[package]]
 name = "datafusion"
-version = "21.0.0"
+version = "21.1.0"
```
Comment (vvv) [inline Cargo.lock:1671]: > fix: panic with datafusion 21

Two questions:
1. What was the problem?
2. How was the problem fixed?

File: Cargo.toml [hunk 2]
```diff
@@ -62,8 +62,8 @@ clap = { version = "4.1.6", default-features = false, features = [
   "cargo",
 ] }
 dashmap = { version = "5.4.0", features = ["serde"] }
-datafusion = { version = "21.0", features = ["simd"] }
-datafusion-common = "21.0"
+datafusion = { version = "21.1.0", features = ["simd"] }
+datafusion-common = "21.1.0"
```
Comment (vvv) [inline Cargo.toml:66]: [optional]
Consider omitting the “patch” version. You'll be able to update `Cargo.lock` with `cargo update -p datafusion` without editing `Cargo.toml`.
```suggestion
datafusion = { version = "21.1", features = ["simd"] }
datafusion-common = "21.1"
```

**Comment by vvv** [comment]
Ah, I'm late to the reviewers' party! 🐌 :)

File: Cargo.toml (hunk 2 above)
Comment (hengfeiyang) [reply Cargo.toml:66]: you are right

File: Cargo.lock (hunk 1 above)
Comment (hengfeiyang) [reply Cargo.lock:1671]: just fixed a version conflict https://github.com/apache/arrow-datafusion/compare/21.0.0...21.1.0

File: src/service/search/mod.rs [hunk 3]
```diff
@@ -678,6 +684,23 @@ pub fn handle_datafusion_error(err: DataFusionError) -> Error {
...
+        if pos_start.is_none() {
+            continue;
+        }
+        let pos_start = pos_start.unwrap();
+        let pos_end = err[pos + pos_start + 1..].find(ponct);
+        if pos_end.is_none() {
+            continue;
+        }
+        let pos_end = pos_end.unwrap();
+        return Some(err[pos + pos_start + 1..pos + pos_start + 1 + pos_end].to_string());
```
[the remark this answers is not in the document]
Comment (hengfeiyang) [reply src/service/search/mod.rs:699]: you reduce one line.

**Pull request merged**